Add experimental dashboard to monitor simulations

nilsvu opened the pull request:
## Proposed changes

This takes the `status` data and displays it in a web browser for all running jobs on the cluster, along with lots of plots to show the current state of BBH simulations. It's experimental, in that to use it you have to `pip install streamlit` by yourself and run `bin/python-spectre -m streamlit run tools/Status/Dashboard.py` to start the dashboard. There are some instructions in the file.

Here's a glance at what it looks like:

<img width="1307" alt="Bildschirmfoto 2024-06-29 um 14 04 37" src="https://github.com/sxs-collaboration/spectre/assets/746230/47d21c7a-f310-44d7-9389-727282a70940">

I found this so useful when monitoring BBH simulations that I decided to open this experimental PR.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Split some CLIs into a free function

Allows to call the function separately from the CLI.

nilsvu committed:
Add experimental dashboard to monitor simulations

nilsvu committed:
Add dashboard for BBH

knelli2 commented on tools/Status/Dashboard.py:
```diff
@@ -0,0 +1,181 @@
+    with outfile_tab:
+        outfile = run_dir / "spectre.out"
+        if outfile.exists():
+            with open(outfile, "r") as open_outfile:
+                st.code("".join(["...\n"] + open_outfile.readlines()[-30:]))
```
Any way to make this scrollable?

nilsvu replied:
I tried loading the full file but that broke the server :D Added a comment

knelli2 commented on tools/Status/Dashboard.py:
```diff
@@ -0,0 +1,181 @@
+
+st.divider()
+
+# Refresh the page regularly
+if st.toggle("Auto-refresh", True):
```
I found that if there is an error rending one of the plots of an executable status, then the auto-refresh is somehow disabled. Clicking on a different run brings it back. Any ideas on what to do for this?

nilsvu replied:
Ah ok I moved the auto refresh further up in the script, that should fix it

knelli2 commented on tools/Status/ExecutableStatus/ExecutableStatus.py:
```diff
@@ -66,6 +66,25 @@ def format(self, field: str, value: Any) -> str:
+
+        Arguments:
+          job: A dictionary of job information, including the input file and
+            working directory. See the 'spectre.tools.Status.status' function
```
`fetch_status` or `render_status` to be explicit?

knelli2 commented on tools/Status/ExecutableStatus/EvolveGhBinaryBlackHole.py:
```diff
@@ -94,3 +97,141 @@ def format(self, field, value):
+                            "Time"
+                        )[
+                            [
+                                "L2Norm(PointwiseL2Norm(GaugeConstraint))",
+                                "L2Norm(PointwiseL2Norm(ThreeIndexConstraint))",
+                                "L2Norm(PointwiseL2Norm(FourIndexConstraint))",
```
Maybe use
```py
[col for col in df.columns if "Constraint" in col]
```
so that it will work if people stray from the pipeline a bit?

knelli2 commented on tools/Status/ExecutableStatus/EvolveGhBinaryBlackHole.py:
```diff
@@ -94,3 +97,141 @@ def format(self, field, value):
+
+        render_control_systems()
+
+        # Eccentricity
+        # (can be enabled once EccentricityControl is available)
+        # st.subheader("Eccentricity")
```
I think you've updated this section since you opened this. Wanna add those changes in?

nilsvu replied:
It's enabled now since #5895 was merged

knelli2 commented on tools/Status/ExecutableStatus/ExecutableStatus.py:
```diff
@@ -190,6 +202,51 @@ def format(self, field, value):
+            input_file, list_reduction_files(job, input_file)
+        )
+
 
 class EllipticStatus(ExecutableStatus):
```
I think you've also added a dashboard for the elliptic one too since you opened this. Wanna add that? I think the only thing is figuring out how to show the latest control step and not just the first one.

nilsvu replied:
I added the elliptic dashboard. It only shows the convergence for the first control step atm, and also plots the convergence of the control parameter. I'll probably combine these somehow, but haven't decided how yet. Also AMR iterations are still missing from the status.

knelli2 reviewed (commented):
These are all optional really. Just some things I noticed. We can't test any of this so once you feel it's ready we can just merge

nilsvu committed:
PlotTrajectories: load time column

nilsvu committed:
Add dashboard for elliptic executables

nilsvu committed:
Minor updates to EccentricityControl.py

- Fix issue with trajectories having different lengths
- Pass in an MPL figure for plots instead of always
  writing the plot to a file.

knelli2 commented on support/Pipelines/EccentricityControl/EccentricityControl.py:
```diff
@@ -120,10 +120,7 @@ def compute_separation(h5_file, subfile_name_aha, subfile_name_ahb):
-    num_obs = len(ObjectA_centers[:, 0])
-
-    if len(ObjectB_centers[:, 0]) < num_obs:
-        num_obs = len(ObjectB_centers[:, 0])
+    num_obs = min(len(ObjectA_centers[:, 0]), len(ObjectB_centers[:, 0]))
```
Maybe this commit and the one before should be in a separate PR since they are a bug fix? That way they don't get buried in an experimental PR

knelli2 reviewed (changes requested):
Everything else looks ok

knelli2 commented:
@nilsvu  once the tests are fixed, we can merge this

nilsvu committed:
Fix isort config

Make sure isort always treats the `spectre` imports as
first party. Before this fix, isort could get confused if the
source directory was not named "spectre".

nilsvu committed:
Add CCE executable status and dashboard

nilsvu commented:
Ok I squashed and rebased. Thanks for reviewing!
